plugin/forward/connect.go

```diff
@@ -87,4 +94,8 @@ func (p *Proxy) connect(ctx context.Context, state request.Request, forceTCP, me
 	return ret, nil
 }
 
-const rttCount = 4
+const (
+	rttCount   = 4
+	minTimeout = time.Millisecond * 500
+	maxTimeout = dialTimeout // 4 s
```

Comment:
I think 4s is too much. Such timeout will eat almost all our queryTimeout time during which we can retry the query. I would suggest 
`maxTimeout = timeout` as here https://github.com/coredns/coredns/blob/d8b9dec4b35932b35b44b17a4b68883fda7c30d0/plugin/forward/proxy.go#L93